pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/groupby/DefaultGroupKeyGenerator.java

```diff
@@ -365,20 +562,89 @@ public boolean hasNext() {
 
     @Override
     public GroupKey next() {
-      Map.Entry<Long, Integer> entry = _iterator.next();
-
-      long groupKey = entry.getKey();
-      int groupId = entry.getValue();
+      Object2IntMap.Entry<IntArrayList> entry = _iterator.next();
 
-      String stringGroupKey = rawKeyToStringGroupKey(groupKey);
-      _groupKey.setFirst(groupId);
+      String stringGroupKey = rawKeyToStringGroupKey(entry.getKey());
+      _groupKey.setFirst(entry.getIntValue());
       _groupKey.setSecond(stringGroupKey);
       return _groupKey;
     }
 
     @Override
     public void remove() {
-      throw new RuntimeException("Unsupported remove method.");
+      throw new UnsupportedOperationException();
     }
   }
+
+  /**
+   * With an integer group key, convert group key from dictId based to string based, using actually values corresponding
+   * to dictionary id's.
+   * (ARRAY_BASED storage type)
+   *
+   * @param groupKey integer group key.
+   * @return string group key.
+   */
+  private String groupKeyToStringGroupKey(int groupKey) {
+    if (_numGroupByColumns == 1) {
+      // Special case one group-by column for performance.
+      return _dictionaries[0].get(groupKey).toString();
+    } else {
+      // Decode the group key.
+      int cardinality = _cardinalities[0];
+      StringBuilder builder = new StringBuilder(_dictionaries[0].get(groupKey % cardinality).toString());
```

Comment:
There is no effecient way to clear the string builder, so I think it's better just get a new one each time (8 byte garbage?)